### DISCUSSION TITLE
StatusLowBattery Characteristics - question

### DISCUSSION THREAD

**Discussion by homonto** [HomeSpan/HomeSpan#442] [General]
hi,

So I created an accessory:
```
  new SpanAccessory();
    new UpdateData(DEVICE_ID_2_NAME,DEVICE_ID_2);                 
    new RemoteTempSensor("Temperature Sensor",DEVICE_ID_2);        
    new RemoteHumSensor("Humidity Sensor",DEVICE_ID_2);        
    new RemoteLightSensor("Light Sensor",DEVICE_ID_2);  
```   
and this is 1 box, that has 3 sensors (temp, hum, light).
Since it is 1 box, I thought I could add "StatusLowBattery" characteristic but apparently Accessory does not accept it (Warning after hub starts) - although I saw the warning, iPhone is showing proper info on the screen: Low battery.
But I don't like warnings, so I went another way: while I receive the data (low battery = 1) in this method:

```
struct UpdateData : Service::AccessoryInformation
{
  SpanCharacteristic *md_version;

  SpanPoint *remoteData;
  const char *name;
  const char *macAddress;
  
  UpdateData(const char *name, const char*macAddress) : Service::AccessoryInformation()
  {
    this->name=name;
    this->macAddress=macAddress;
    new Characteristic::Identify();
    new Characteristic::Name(name);

    md_version  = new Characteristic::FirmwareRevision(DEVICE_FW);
    remoteData  = new SpanPoint(macAddress,0,sizeof(myData));    // create a SpanPoint with send size=0 and receive size=sizeof(myData)
  }

...
  void loop()
  {                                       // if there is data from the remote sensor
    if(remoteData->get(&myData))
    {  
      low_bat_value         = myData.md_low_bat;  // update low battery
...
```

I pass it to the next struct and method:
```
struct RemoteTempSensor : Service::TemperatureSensor
{
  SpanCharacteristic *temp;
  SpanCharacteristic *fault;
  SpanCharacteristic *low_bat;
  

  const char *name;
  const char *macAddress;

  RemoteTempSensor(const char *name, const char*macAddress) : Service::TemperatureSensor()
  {
    this->name=name;
    this->macAddress=macAddress;
    temp=new Characteristic::CurrentTemperature(0);           // set initial temperature
    temp->setRange(-50,100);                                  // expand temperature range to allow negative values
    fault=new Characteristic::StatusFault(0);                 // set initial state = no fault
    low_bat=new Characteristic::StatusLowBattery(1);          // initial battery status, 1=low battery, 0=battery OK
  }

```
then I update the value:
`low_bat->setVal(low_bat_value);   `     
in the Service::TemperatureSensor()

Now, I don't have warning, but crazy thing is: all 3 sensors in this accessory (temp, hum and light) have this property so it is enough to update setVal(low_bat_value) on 1 of the sensors and all get the warning on the iPhone

Is that like HomeKit works, or your library works?

**Comment by HomeSpan** [comment]
If you are only adding the StatusLowBattery Characteristic to the Temperature Sensor Service, but it appears on each of the three Services (temp, hum, light) in the Home App, this must be a "feature" of Home Kit itself, or more likely the Home App.  To double check, see how Eve for HomeKit displays your device.  Eve tends to be more faithful and displays Services exactly as they are provided, whereas with each new update to iOS, Apple seems to aggregate and bundle things differently in the Home App.

**Reply by homonto** [reply]
thank you - I will check it. However, it is kind of logical: if 3 sensors are in 1 accessory I would expect them to have the same battery ;-)
More, I would even allow to add battery to the accessory itself rather than to each sensor
but what do I know ;-)

**Comment by homonto** [comment]
Ha, so Eve is also crazy: once it shows in Temperature Sensor only, another view it shows on All:
![IMG_0011](https://user-images.githubusercontent.com/46562447/202996116-cbac1ed0-d748-4f83-84db-334e24c78d10.jpeg)
![IMG_0012](https://user-images.githubusercontent.com/46562447/202996129-ada490cf-021f-44b9-b093-164876c389be.jpeg)

**Comment by homonto** [comment]
I have rewritten this code and I don't use  StatusLowBattery on Temperature Sensor itself anymore but on the box so the battery is now common for all sensors connected to this box (so humidity, temperature, light etc.):

```
    new SpanAccessory();
      new UpdateData(DEVICE_ID_1_NAME,DEVICE_ID_1);                 
      new RemoteTempSensor("Temperature Sensor",DEVICE_ID_1);        
      new RemoteHumSensor("Humidity Sensor",DEVICE_ID_1);        
      new RemoteLightSensor("Light Sensor",DEVICE_ID_1);  
      new RemoteBattery(DEVICE_ID_1_NAME,DEVICE_ID_1);  
```  

and inside 
`RemoteBattery(DEVICE_ID_1_NAME,DEVICE_ID_1);    `
I do:

```
struct RemoteBattery : Service::BatteryService
{
  SpanCharacteristic *battery_level;
  SpanCharacteristic *charging_state;
  SpanCharacteristic *low_battery;
  const char *name;
  const char *macAddress;

  RemoteBattery(const char *name, const char*macAddress) : Service::BatteryService()
  {
    this->name=name;
    this->macAddress=macAddress;
    new Characteristic::Name("Battery sensor");
    battery_level     = new Characteristic::BatteryLevel(0);                     // set initial state to 0%
    charging_state    = new Characteristic::ChargingState(0);                    // set initial state to "not charging"
    low_battery       = new Characteristic::StatusLowBattery(1);                 // set initial state to true
  }

  void loop()
  {
    if (bat_received)
    {
      battery_level->setVal(bat_value_pct);   
      if (bat_value_pct < 30) 
      {
        low_bat=1;
      } else
      {
        low_bat=0;
      }
      low_battery->setVal(low_bat);  
      bat_received = false;

      if (charging_received)
      {
        if (md_charging_value > 0)
          charging_state->setVal(1);
        else 
          charging_state->setVal(0);
      }
      LOG1("[%s]: Battery data from sensor %s on device %s: battery level=%d%%, low_battery=%d, charging=%d\n",__func__,name,macAddress,bat_value_pct,low_bat,md_charging_value);
      
```
